### PR TITLE
Remove duplicates solution

### LINKED ISSUE #9: Remove Duplicates from Sorted List
### Problem Description:
You are given the head of a sorted linked list. Remove all duplicates so that each element appears only once. Return the linked list sorted as well.

### Expected Behavior:
- Input: head = [1,1,2]  
  Output: [1,2]

- Input: head = [1,1,2,3,3]  
  Output: [1,2,3]

- Input: head = []  
  Output: []

### Constraints:
- The number of nodes in the list is in the range [0, 300].
- -100 <= Node.val <= 100.
- The list is guaranteed to be sorted in ascending order.

### Link to Challenge:
[Remove Duplicates from Sorted List](https://leetcode.com/problems/remove-duplicates-from-sorted-list/description/)

### PR DESCRIPTION
**Pull request by Shahroz657**
fixes #9 
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior
The function removes duplicate values from a sorted linked list, ensuring each value appears only once while preserving the list's original sorted order. It traverses the list, adjusting node pointers to skip duplicate nodes.

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/remove_duplicates_from_sorted_list.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    Parameters:
+        head (Optional[ListNode]): The head node of the sorted linked list.
+
+    Returns:
+        Optional[ListNode]: The head node of the modified linked list with duplicates removed.
+
+    Raises:
+        AssertionError: If the argument is not of type ListNode or None.
+
+    Examples:
```
Comment (fatima-malik99) [inline solutions/remove_duplicates_from_sorted_list.py:36]: Your docstring and doctests for the remove_duplicates function are well-written and clear. However, as per the checklist requirement, there should be at least three passing doctests, but currently, there are only two.
I suggest adding another doctest. For example, you could test a scenario where the input list has no duplicates or is empty.

File: solutions/remove_duplicates_from_sorted_list.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        >>> head = ListNode(1, ListNode(1, ListNode(2)))
+        >>> result = remove_duplicates(head)
+        >>> result.val, result.next.val, result.next.next is None
+        (1, 2, True)
+
+        >>> head = ListNode(1, ListNode(1, ListNode(2, ListNode(3, ListNode(3)))))
+        >>> result = remove_duplicates(head)
+        >>> result.val, result.next.val, result.next.next.val
+        (1, 2, 3)
```
Comment (fatima-malik99) [inline solutions/remove_duplicates_from_sorted_list.py]: Your docstring and doctests for the remove_duplicates function are well-written and clear. However, as per the checklist requirement, there should be at least three passing doctests, but currently, there are only two.
I suggest adding another doctest. For example, you could test a scenario where the input list has no duplicates or is empty.

File: solutions/remove_duplicates_from_sorted_list.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+class ListNode:
+    """A node in a singly linked list."""
+
+    def __init__(self, val=0, next_node=None):
+        self.val = val
+        self.next = next_node
+
+
+def remove_duplicates(head: Optional[ListNode]) -> Optional[ListNode]:
```
Comment (fatima-malik99) [inline solutions/remove_duplicates_from_sorted_list.py]: To follow Evan's guideline, it would be great if you rename the function to remove_duplicates_from_sorted_list to match the file name remove_duplicates_from_sorted_list.py. Otherwise, everything else looks good!

File: solutions/tests/test_remove_duplicates_from_sorted_list.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        result = remove_duplicates(head)
+        self.assertEqual(result.val, 1)
+
+    def test_single_node_next(self):
+        """It should evaluate the next node of a single-node list as None."""
+        head = ListNode(1)
+        result = remove_duplicates(head)
+        self.assertIsNone(result.next)
+
+    def test_duplicate_first_node(self):
```
Comment (fatima-malik99) [inline solutions/tests/test_remove_duplicates_from_sorted_list.py:40]: Your code is very clean but I have a small suggestion regarding the test case names and docstrings. While they are generally clear a few of them could be slightly more descriptive to improve readability.
Like in this test, instead of test_duplicate_first_node, a name like test_remove_duplicate_at_head would make it clearer that the duplicate appears at the beginning of the list.

File: solutions/tests/test_remove_duplicates_from_sorted_list.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        result = remove_duplicates(head)
+        self.assertEqual(result.val, 1)
+
+    def test_duplicate_second_node(self):
+        """It should remove duplicates at the second node."""
+        head = ListNode(1, ListNode(1, ListNode(2)))
+        result = remove_duplicates(head)
+        self.assertEqual(result.next.val, 2)
+
+    def test_duplicate_list_ends(self):
```
Comment (fatima-malik99) [inline solutions/tests/test_remove_duplicates_from_sorted_list.py:52]: Similarly, test_duplicate_list_ends could be renamed to test_remove_duplicates_at_tail to specify that duplicates appear at the end.

File: solutions/tests/test_remove_duplicates_from_sorted_list.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        """It should evaluate the end of a list with duplicates as None."""
```
Comment (fatima-malik99) [inline solutions/tests/test_remove_duplicates_from_sorted_list.py]: It's a good docstring but if I have to look from a beginner's point of view so it can be write more simply. "It should remove duplicates at the end, ensuring the last node is correct."
This makes it clearer that the focus is on handling duplicates at the end while ensuring the last node remains valid. It avoids any ambiguity.

File: solutions/tests/test_remove_duplicates_from_sorted_list.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        self.assertEqual(result.next.val, 2)
+
+    def test_list_without_duplicates_third_node(self):
+        """It should evaluate the third node of a list without duplicates."""
+        head = ListNode(1, ListNode(2, ListNode(3)))
+        result = remove_duplicates(head)
+        self.assertEqual(result.next.next.val, 3)
+
+    def test_list_without_duplicates_end(self):
+        """It should evaluate the end of a list without duplicates as None."""
```
Comment (fatima-malik99) [inline solutions/tests/test_remove_duplicates_from_sorted_list.py]: Your docstring is good, but if you write it like this instead →
"It should keep the last node unchanged in a list without duplicates."
This change helps make it clearer that the list should stay the same if there are no duplicates. However, the wording might make it seem like the last node should be None, which isn’t the case. A small change could make it even clearer. 🙂

File: solutions/tests/test_remove_duplicates_from_sorted_list.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    Unit tests for the remove_duplicates function.
+    Ensures that duplicates are removed from a sorted linked list.
+    """
+
+    def test_empty_list(self):
+        """It should evaluate None to None."""
+        result = remove_duplicates(None)
+        self.assertIsNone(result)
+
+    def test_single_node_value(self):
```
Comment (fatima-malik99) [inline solutions/tests/test_remove_duplicates_from_sorted_list.py:28]: Nice work on covering the basic boundary cases like the empty list and single-node list! These are important to check. If you'd like, you could also consider adding a few more boundary cases, such as:

A list with all identical elements (e.g., ListNode(1, ListNode(1, ListNode(1)))).
A list with alternating duplicates (e.g., ListNode(1, ListNode(1, ListNode(2, ListNode(2))))).
Testing with a larger list (e.g., 1000 nodes) to check performance. 
Adding these cases is optional, but I believe they would make your tests even more good. Feel free to include them if you think it’s necessary. I would love to hear your point of view as well.

File: solutions/remove_duplicates_from_sorted_list.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        >>> head = ListNode(1, ListNode(1, ListNode(2)))
+        >>> result = remove_duplicates(head)
+        >>> result.val, result.next.val, result.next.next is None
+        (1, 2, True)
+
+        >>> head = ListNode(1, ListNode(1, ListNode(2, ListNode(3, ListNode(3)))))
+        >>> result = remove_duplicates(head)
+        >>> result.val, result.next.val, result.next.next.val
+        (1, 2, 3)
+    """
```
Comment (fatima-malik99) [inline solutions/remove_duplicates_from_sorted_list.py:57]: Everything is great, you did an amazing job but there is one thing to note that the function assumes the input list is sorted in non-decreasing order, as it relies on comparing consecutive nodes to detect duplicates. This assumption, however, isn't currently documented in the function's docstring. According to the code review checklist, we have to document other assumptions as well.

**Comment by MadiMalik** [comment]
What will happen if the list is not sorted?

File: solutions/remove_duplicates_from_sorted_list.py (hunk 2 above)
Comment (Shahroz657) [reply solutions/remove_duplicates_from_sorted_list.py]: Hi Fatima,

Thank you for the feedback and for pointing out the missing doctests. I have updated the function to include additional doctests to meet the checklist requirements. Specifically:
1.  Added a doctest for an input list with no duplicates.
2.  A doctest for an empty input list (None) was added.
Now, the function includes four doctests that cover different scenarios, including the ones you mentioned.
Thank you for your valuable insights!

File: solutions/remove_duplicates_from_sorted_list.py (hunk 3 above)
Comment (Shahroz657) [reply solutions/remove_duplicates_from_sorted_list.py]: Thank you for pointing that out! You're absolutely correct—renaming the function to match the file name would align better with the guidelines and improve consistency. I will update the function name to remove_duplicates_from_sorted_list to ensure it adheres to Evan's guidelines.

I appreciate your attention to detail and your constructive feedback!

File: solutions/tests/test_remove_duplicates_from_sorted_list.py (hunk 4 above)
Comment (Shahroz657) [reply solutions/tests/test_remove_duplicates_from_sorted_list.py:40]: Hi Fatima,

Thank you for the suggestion and for taking the time to review the code! I agree that clear and descriptive names for test cases are important for readability. However, in this case, I believe the current name, test_duplicate_first_node, adequately conveys the intent of the test—that the duplicate is located at the first node.

While test_remove_duplicate_at_head is a good alternative, I feel the current naming is still clear and concise. If there are further concerns about clarity, I'd be happy to revisit this with more input.

Thanks again for your valuable feedback!

File: solutions/tests/test_remove_duplicates_from_sorted_list.py (hunk 5 above)
Comment (Shahroz657) [reply solutions/tests/test_remove_duplicates_from_sorted_list.py:52]: Hi Fatima,

Thank you for your suggestion and attention to detail! While test_remove_duplicates_at_tail is a good alternative, I feel the current name, test_duplicate_list_ends, effectively communicates the intent of the test. It indicates that the test is verifying behavior when duplicates are at the ends of the list.

That said, if there’s broader consensus or additional feedback favoring the alternative name, I’m open to revisiting this. I appreciate your thoughtful review!

Best regards,
Shahroz

File: solutions/tests/test_remove_duplicates_from_sorted_list.py (hunk 6 above)
Comment (Shahroz657) [reply solutions/tests/test_remove_duplicates_from_sorted_list.py]: Thank you for your thoughtful suggestion! I understand the importance of making docstrings accessible to beginners, but I believe the current phrasing of the docstring is already clear and concise. It effectively conveys the purpose of the test without introducing ambiguity.

That said, I’ll keep your feedback in mind for future cases where simplification might be more beneficial.

File: solutions/tests/test_remove_duplicates_from_sorted_list.py (hunk 7 above)
Comment (Shahroz657) [reply solutions/tests/test_remove_duplicates_from_sorted_list.py]: I see your point about how the phrasing could be adjusted for clarity. However, I believe the current wording of the docstring is both accurate and clear in describing the purpose of the test. It avoids ambiguity and aligns well with the overall style of the code.

That being said, I appreciate your attention to detail and your effort to make improvements—it’s always great to have fresh perspectives! I’ll definitely keep this feedback in mind for future updates where such refinements might be helpful.

File: solutions/remove_duplicates_from_sorted_list.py (hunk 9 above)
Comment (Shahroz657) [reply solutions/remove_duplicates_from_sorted_list.py:57]: Thank you for pointing this out! I’ve updated the function docstring to explicitly document the assumption that the input list is sorted in non-decreasing order. I appreciate the detailed feedback—it makes the code and documentation even better!

Comment (fatima-malik99) [reply solutions/remove_duplicates_from_sorted_list.py:57]: That's perfect now.

File: solutions/tests/test_remove_duplicates_from_sorted_list.py (hunk 4 above)
Comment (fatima-malik99) [reply solutions/tests/test_remove_duplicates_from_sorted_list.py:40]: That makes sense. I just suggested an alternative, but I see your point. Thanks for considering my feedback :)

File: solutions/tests/test_remove_duplicates_from_sorted_list.py (hunk 5 above)
Comment (fatima-malik99) [reply solutions/tests/test_remove_duplicates_from_sorted_list.py:52]: Well yes, I agree with your point. Thanks for considering my suggestion.

File: solutions/tests/test_remove_duplicates_from_sorted_list.py (hunk 7 above)
Comment (fatima-malik99) [reply solutions/tests/test_remove_duplicates_from_sorted_list.py]: I appreciate your perspective. We can keep it like the way it is now.

File: solutions/tests/test_remove_duplicates_from_sorted_list.py (hunk 6 above)
Comment (fatima-malik99) [reply solutions/tests/test_remove_duplicates_from_sorted_list.py]: Thank you so much for considering my feedback. I am still learning and I believe that your perspective is also good.

File: solutions/remove_duplicates_from_sorted_list.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/remove_duplicates_from_sorted_list.py]: Great Shohroz, you can change the name of the function then I can check this point in code checklist. We are very close to merging this branch.

Comment (Shahroz657) [reply solutions/remove_duplicates_from_sorted_list.py]: Name is already changed, Please check

File: solutions/remove_duplicates_from_sorted_list.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    assert head is None or isinstance(
+        head, ListNode
+    ), "Input must be a ListNode or None."
+
+    current_node = head
+    while current_node and current_node.next:
+        if current_node.value == current_node.next.value:
+            current_node.next = current_node.next.next  # Remove duplicate
+        else:
+            current_node = current_node.next  # Move to next node
```
Comment (fatima-malik99) [inline solutions/remove_duplicates_from_sorted_list.py]: According to the code review checklist, comments should focus on explaining the strategy rather than the implementation. Your current comments describe what each line does, but adding a high-level comment about the approach like 'Traverse the list and remove consecutive duplicates by adjusting pointers' could make it even clearer. Just a small suggestion to align with the guidelines

Comment (Shahroz657) [reply solutions/remove_duplicates_from_sorted_list.py]: Great suggestion

File: solutions/remove_duplicates_from_sorted_list.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/remove_duplicates_from_sorted_list.py]: But I can see that your function name is still not exactly matching with the file name. Please check here 

def remove_duplicates(head: Optional[ListNode]) -> Optional[ListNode]:

File: solutions/remove_duplicates_from_sorted_list.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+class ListNode:
+    """A node in a singly linked list."""
+
+    def __init__(self, value=0, next_node=None):
+        self.value = value
+        self.next = next_node
+
+
+def remove_duplicates_from_sorted_list(head: Optional[ListNode]) -> Optional[ListNode]:
```
Comment (Shahroz657) [inline solutions/remove_duplicates_from_sorted_list.py:23]: @fatima-malik99 Please check. it has already been changed.

File: solutions/remove_duplicates_from_sorted_list.py (hunk 3 above)
Comment (Shahroz657) [reply solutions/remove_duplicates_from_sorted_list.py]: It has been changed. please look into the files changed. Currently, you are looking at the outdated file.

File: solutions/remove_duplicates_from_sorted_list.py (hunk 11 above)
Comment (fatima-malik99) [reply solutions/remove_duplicates_from_sorted_list.py:23]: That's great, now your function name is exactly matching with the file name.

File: solutions/remove_duplicates_from_sorted_list.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/remove_duplicates_from_sorted_list.py]: Yes, it's changed.

**Comment by Shahroz657** [comment]
> What will happen if the list is not sorted?

The function assumes that the input list is sorted in non-decreasing order, as mentioned in the docstring:

> ```python
> Parameters:
>     head (Optional[ListNode]): The head node of the sorted linked list.
>         - The list is assumed to be sorted in non-decreasing order.
> ```

If the input list is not sorted, the behavior would not be as intended because the function relies on consecutive node comparisons to detect duplicates. It will also raise an assertion error for an invalid input.

**Review by fatima-malik99 — APPROVED** [review]
Hey Shahroz, Nice job on this. Finally we can merge this branch. I really appreciate the efforts that you have put in this solution.   Great job.

File: solutions/remove_duplicates_from_sorted_list.py (hunk 2 above)
Comment (fatima-malik99) [reply solutions/remove_duplicates_from_sorted_list.py]: I have checked the new doctests, they are running and now one more check is pass.
